praw/reddit.py

```diff
@@ -464,29 +465,39 @@ def info(self, fullnames=None, url=None):
                   different set of submissions.
 
         """
-        if bool(fullnames) == bool(url):
+        none_count = [fullnames, url].count(None)
+        if none_count > 1:
             raise TypeError("Either `fullnames` or `url` must be provided.")
-        if fullnames:
-            if not isinstance(fullnames, list):
-                raise TypeError("fullnames must be a list")
-
-            def generator():
-                for position in range(0, len(fullnames), 100):
-                    fullname_chunk = fullnames[position : position + 100]
-                    params = {"id": ",".join(fullname_chunk)}
+        if none_count < 1:
+            raise TypeError(
+                "Mutually exclusive parameters: `fullnames`, `url`"
+            )
+
+        if fullnames is not None:
+            if isinstance(fullnames, string_types):
+                raise TypeError("`fullnames` must be a non-str iterable.")
+
+            def generator(fullnames):
+                iterable = iter(fullnames)
+                while True:
+                    chunk = list(islice(iterable, 100))
+                    if not chunk:
+                        break
+
+                    params = {"id": ",".join(chunk)}
                     for result in self.get(API_PATH["info"], params=params):
                         yield result
 
-            return generator()
+            return generator(fullnames)
 
-        try:
-            params = {"url": url}
-            url_list = [
-                result for result in self.get(API_PATH["info"], params=params)
-            ]
-            return url_list
-        except Exception:
-            raise TypeError("Invalid URL or no posts exist")
+        if url is not None:
```

Comment:
Is this condition necessary? It seems like it'll always be true at this point.

Reply:
Yes, I’ve removed it.

A [singledispatch][1] (introduced in Python 3.4) would be a fun thing to use here to separate the logic of the two…

[1]: https://docs.python.org/3/library/functools.html#functools.singledispatch

It made me think though, might it be better if the url parameter were a separate `.info_url()` method?

Reply:
Yeah I think separating the methods probably makes sense. A lot of the functions closely mimicked Reddit's API previously. But I think submission was the first one to get out-of-hand with all different types of submissions, that making specific functions for each makes sense.